### ISSUE TITLE
lxc image copy fails to copy aliases and set the new one - when image type changes

### ISSUE THREAD

**Issue by cr1st1p** [lxc/lxd#6419]
# Required information

 * Distribution: archlinux (virtual machine)
 * Distribution version: 2019-10
 * The output of "lxc info" or if that fails:
```

config: {}
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
- container_mount_propagation
- container_backup
- devlxd_images
- container_local_cross_pool_handling
- proxy_unix
- proxy_udp
- clustering_join
- proxy_tcp_udp_multi_port_handling
- network_state
- proxy_unix_dac_properties
- container_protection_delete
- unix_priv_drop
- pprof_http
- proxy_haproxy_protocol
- network_hwaddr
- proxy_nat
- network_nat_order
- container_full
- candid_authentication
- backup_compression
- candid_config
- nvidia_runtime_config
- storage_api_volume_snapshots
- storage_unmapped
- projects
- candid_config_key
- network_vxlan_ttl
- container_incremental_copy
- usb_optional_vendorid
- snapshot_scheduling
- container_copy_project
- clustering_server_address
- clustering_image_replication
- container_protection_shift
- snapshot_expiry
- container_backup_override_pool
- snapshot_expiry_creation
- network_leases_location
- resources_cpu_socket
- resources_gpu
- resources_numa
- kernel_features
- id_map_current
- event_location
- storage_api_remote_volume_snapshots
- network_nat_address
- container_nic_routes
- rbac
- cluster_internal_copy
- seccomp_notify
- lxc_features
- container_nic_ipvlan
- network_vlan_sriov
- storage_cephfs
- container_nic_ipfilter
- resources_v2
- container_exec_user_group_cwd
- container_syscall_intercept
- container_disk_shift
- storage_shifted
- resources_infiniband
- daemon_storage
- instances
- image_types
- resources_disk_sata
- clustering_roles
- images_expiry
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    REDACTED
    -----END CERTIFICATE-----
  certificate_fingerprint: REDACTED
  driver: lxc
  driver_version: 3.2.1
  kernel: Linux
  kernel_architecture: x86_64
  kernel_features:
    netnsid_getifaddrs: "true"
    seccomp_listener: "true"
    shiftfs: "false"
    uevent_injection: "true"
    unpriv_fscaps: "true"
  kernel_version: 5.3.1-arch1-1-ARCH
  lxc_features:
    mount_injection_file: "true"
    network_gateway_device_route: "true"
    network_ipvlan: "true"
    network_l2proxy: "true"
    network_phys_macvlan_mtu: "true"
    seccomp_notify: "true"
  project: default
  server: lxd
  server_clustered: false
  server_name: 1162-arch-2019-10
  server_pid: 56234
  server_version: "3.18"
  storage: dir
  storage_version: "1"


```

# Issue description
I want to bring locally an image, with all its aliases and it fails setting **any** alias:
```
lxc image copy ubuntu:18.04 local: --copy-aliases --auto-update --alias ubuntu-18.04 

Image copied successfully!                   
Failed to create alias ubuntu-18.04
Failed to create alias 18.04
Failed to create alias 18.04/amd64
Failed to create alias b
Failed to create alias b/amd64
Failed to create alias bionic
Failed to create alias bionic/amd64
Failed to create alias default
Failed to create alias default/amd64
Failed to create alias lts
Failed to create alias lts/amd64
Failed to create alias ubuntu
Failed to create alias amd64

lxc image list
+-------+--------------+--------+---------------------------------------------+--------+-----------------+----------+-----------------------------+
| ALIAS | FINGERPRINT  | PUBLIC |                 DESCRIPTION                 |  ARCH  |      TYPE       |   SIZE   |         UPLOAD DATE         |
+-------+--------------+--------+---------------------------------------------+--------+-----------------+----------+-----------------------------+
|       | c3f79185497f | no     | ubuntu 18.04 LTS amd64 (release) (20191021) | x86_64 | VIRTUAL-MACHINE | 328.44MB | Nov 7, 2019 at 7:46pm (UTC) |
+-------+--------------+--------+---------------------------------------------+--------+-----------------+----------+-----------------------------+

```

What I think happens, is that the image gets converted to a different type, which means a different fingerprint, but at the end of the copy operation, the aliases are tried to be applied on the **original** fingerprint - which is probably wrong:

```
lxc image info ubuntu:18.04 | grep -E 'Fingerprint|Type'
Fingerprint: d6f281a2e523674bcd9822f3f61be337c51828fb0dc94c8a200ab216d12a0fff
Type: container

lxc image --debug  copy ubuntu:18.04 local: --copy-aliases --auto-update --alias ubuntu-18.04
....
DBUG[11-07|19:50:19] Sending request to LXD                   method=POST url=http://unix.socket/1.0/images/aliases etag=
DBUG[11-07|19:50:19] 
        {
                "description": "",
                "target": "d6f281a2e523674bcd9822f3f61be337c51828fb0dc94c8a200ab216d12a0fff",
                "name": "ubuntu-18.04",
                "type": ""
        } 
Failed to create alias ubuntu-18.04
... (and so on)
```
but:
```
 lxc image info c3f79185497f | grep -E 'Fingerprint|Type'
Fingerprint: c3f79185497fb3feae2aa701114637b3c402f166313bc30bb21c87a2b53baead
Type: virtual-machine
```

It might be related to the latest changes regarding #6205
On my host machine, based on Arch, the image was copied ok, and it has type 'container'.

# Steps to reproduce
See above. Note that I was inside a virtual machine, created via vagrant and virtualbox

# Information to attach

 - [ ] Any relevant kernel output (`dmesg`)
 - [ ] Container log (`lxc info NAME --show-log`)
 - [ ] Container configuration (`lxc config show NAME --expanded`)
 - [ ] Main daemon log (at /var/log/lxd/lxd.log or /var/snap/lxd/common/lxd/logs/lxd.log)
 - [x] Output of the client with --debug
 - [ ] Output of the daemon with --debug (alternatively output of `lxc monitor` while reproducing the issue)

**Comment by stgraber** [comment]
Can you run `lxc monitor` in a terminal, then run the copy with `--debug` passed to `lxc image copy` and attach the output of both commands?

I can't seem to reproduce that issue so we're going to have to compare exactly what's happening on both sides to figure it out. `lxc image copy` shouldn't copy a VM image unless `--vm` is passed or it's provided with something which can only be found as a VM image.

**Comment by cr1st1p** [comment]
UPDATE: Adding '--vm' to the lxc image copy command works.
Which suggests to me that there could be actually 2 bugs:
- handling aliases when converting image type - what I reported first
- selecting which image to copy, in absence of '--vm' parameter ? Because I don't think I wanted the 'virtual-machine' one.

Ah, just tested as well:
```
lxc launch c3f79185497f t1
Creating t1
Error: Requested image doesn't match instance type
```
I had to use the fingerprint to have it working:
```
lxc image copy ubuntu:d6f281a2e523  local: --copy-aliases --auto-update --alias ubuntu-18.04 
Image copied successfully!                   
[vagrant@1162-arch-2019-10 ~]$ lxc image list
+-------------+--------------+--------+---------------------------------------------+--------+-----------+----------+-----------------------------+
|    ALIAS    | FINGERPRINT  | PUBLIC |                 DESCRIPTION                 |  ARCH  |   TYPE    |   SIZE   |         UPLOAD DATE         |
+-------------+--------------+--------+---------------------------------------------+--------+-----------+----------+-----------------------------+
| b (12 more) | d6f281a2e523 | no     | ubuntu 18.04 LTS amd64 (release) (20191021) | x86_64 | CONTAINER | 178.05MB | Nov 7, 2019 at 8:13pm (UTC) |
+-------------+--------------+--------+---------------------------------------------+--------+-----------+----------+-----------------------------+
[vagrant@1162-arch-2019-10 ~]$ lxc launch d6f281a2e523 t1
Creating t1
Starting t1

```

You can run the following to see the images related to this issue:
```
 lxc image list ubuntu: serial=20191021 release=bionic architecture=amd64

+-------------+--------------+--------+---------------------------------------------+--------+-----------------+----------+-------------------------------+
|    ALIAS    | FINGERPRINT  | PUBLIC |                 DESCRIPTION                 |  ARCH  |      TYPE       |   SIZE   |          UPLOAD DATE          |
+-------------+--------------+--------+---------------------------------------------+--------+-----------------+----------+-------------------------------+
| b (11 more) | c3f79185497f | yes    | ubuntu 18.04 LTS amd64 (release) (20191021) | x86_64 | VIRTUAL-MACHINE | 328.44MB | Oct 21, 2019 at 12:00am (UTC) |
+-------------+--------------+--------+---------------------------------------------+--------+-----------------+----------+-------------------------------+
| b (11 more) | d6f281a2e523 | yes    | ubuntu 18.04 LTS amd64 (release) (20191021) | x86_64 | CONTAINER       | 178.05MB | Oct 21, 2019 at 12:00am (UTC) |
+-------------+--------------+--------+---------------------------------------------+--------+-----------------+----------+-------------------------------+
|             | 83053a96eaa2 | yes    | ubuntu 18.04 LTS amd64 (release) (20191021) | x86_64 | CONTAINER       | 141.35MB | Oct 21, 2019 at 12:00am (UTC) |
+-------------+--------------+--------+---------------------------------------------+--------+-----------------+----------+-------------------------------+

```

**Comment by cr1st1p** [comment]
@stgraber  will do, but something is strange:
After I did the commands mentioned in the 'UPDATE:' comment, I deleted the image - it was the one brought up via the 'd6' fingerprint. Then did lxc monitor + lxc image copy.... and it worked?!?

I'll recreate the virtual environment :-(

**Comment by stgraber** [comment]
Always annoying when that happens :)

**Comment by cr1st1p** [comment]
Recreated the virtual machine 2 times, but this time they brought the image correctly.
Note that when I created this issue, I think it was my 2nd virtual machine that had the problem.

I found the debug output when lxc asked for the image copy, in my console's buffers, in case that helps:
```
DBUG[11-07|19:50:19] Connected to the websocket 
DBUG[11-07|19:50:19] Sending request to LXD                   method=POST url=http://unix.socket/1.0/images etag=
DBUG[11-07|19:50:19] 
        {
                "auto_update": true,
                "properties": null,
                "public": false,
                "expires_at": "0001-01-01T00:00:00Z",
                "filename": "",
                "source": {
                        "alias": "",
                        "certificate": "",
                        "protocol": "simplestreams",
                        "server": "https://cloud-images.ubuntu.com/releases",
                        "image_type": "",
                        "mode": "pull",
                        "type": "image",
                        "url": "",
                        "name": "",
                        "fingerprint": "18.04",
                        "secret": ""
                },
                "compression_algorithm": "",
                "aliases": null
        } 
DBUG[11-07|19:50:19] Got operation from LXD 
DBUG[11-07|19:50:19] 
        {
                "id": "b1f87d20-3ad8-41f9-82fa-2f3245c384d2",
                "class": "task",
                "description": "Downloading image",
                "created_at": "2019-11-07T19:50:19.262161411Z",
                "updated_at": "2019-11-07T19:50:19.262161411Z",
                "status": "Running",
                "status_code": 103,
                "resources": null,
                "metadata": null,
                "may_cancel": false,
                "err": "",
                "location": "none"
        } 
DBUG[11-07|19:50:19] Sending request to LXD                   method=GET url=http://unix.socket/1.0/operations/b1f87d20-3ad8-41f9-82fa-2f3245c384d2 etag=
DBUG[11-07|19:50:19] Got response struct from LXD 
DBUG[11-07|19:50:19] 
        {
                "id": "b1f87d20-3ad8-41f9-82fa-2f3245c384d2",
                "class": "task",
                "description": "Downloading image",
                "created_at": "2019-11-07T19:50:19.262161411Z",
                "updated_at": "2019-11-07T19:50:19.262161411Z",
                "status": "Running",
                "status_code": 103,
                "resources": null,
                "metadata": null,
                "may_cancel": false,
                "err": "",
                "location": "none"
        } 
Image copied successfully!
DBUG[11-07|19:50:19] Sending request to LXD                   method=GET url="http://unix.socket/1.0/images/aliases?recursion=1" etag=
DBUG[11-07|19:50:19] Got response struct from LXD 
DBUG[11-07|19:50:19] 
        [] 
DBUG[11-07|19:50:19] Sending request to LXD                   method=POST url=http://unix.socket/1.0/images/aliases etag=
DBUG[11-07|19:50:19] 
        {
                "description": "",
                "target": "d6f281a2e523674bcd9822f3f61be337c51828fb0dc94c8a200ab216d12a0fff",
                "name": "ubuntu-18.04",
                "type": ""
        } 
Failed to create alias ubuntu-18.04
DBUG[11-07|19:50:19] Sending request to LXD                   method=POST url=http://unix.socket/1.0/images/aliases etag=
DBUG[11-07|19:50:19] 
        {
                "description": "",
                "target": "d6f281a2e523674bcd9822f3f61be337c51828fb0dc94c8a200ab216d12a0fff",
                "name": "18.04",
                "type": ""
        } 
Failed to create alias 18.04
DBUG[11-07|19:50:19] Sending request to LXD                   method=POST url=http://unix.socket/1.0/images/aliases etag=
DBUG[11-07|19:50:19] 
        {
                "description": "",
                "target": "d6f281a2e523674bcd9822f3f61be337c51828fb0dc94c8a200ab216d12a0fff",
                "name": "18.04/amd64",
                "type": ""
        } 
Failed to create alias 18.04/amd64

```

**Comment by cr1st1p** [comment]
And.. can't stand to not say it - many many thanks for LXD! Using it for a long time and it made my dev work waaaay easier!

**Comment by cr1st1p** [comment]
Looking at the lxc client code, and especially since we're speaking about a virtual machine - think network connection problems, I see this code that rings a bell to me:
client/simplestreams_images.go:
``
// GetImageAlias returns an existing alias as an ImageAliasesEntry struct
func (r *ProtocolSimpleStreams) GetImageAlias(name string) (*api.ImageAliasesEntry, string, error) {
        alias, err := r.ssClient.GetAlias("container", name)
        if err != nil {
                alias, err = r.ssClient.GetAlias("virtual-machine", name)
                if err != nil {
                        return nil, "", err
                }
        }

        return alias, "", err
}
```
So, in case of a timeout, let's say, can I expect that the code will blindly try again by asking for a 'virtual-machine' type alias, and maybe now network is ok, and it actually returns the data ?

Maybe a way more concise error check should be done - like 'if error is NotFound' ...

**Comment by stgraber** [comment]
The request to LXD matches what I'd have expected, so would have been interesting to see what LXD then did internally.

For simplestreams, that's not how this works, multiple calls to `GetAlias` do not result in multiple files being retrieved over the network. We're dealing with a single (and rather large) JSON document which contains all the information for the image server, all the commands just extract data from that JSON document. If there had been a network error, no images would have been reported at all.

**Comment by stgraber** [comment]
@cr1st1p any luck reproducing this?

**Comment by cr1st1p** [comment]
I had a few VM rebuilds but the problem did not manifest again.
Maybe there was a temporarily glitch in the json downloaded from ubuntu, by the lxd daemon.

One more observation, if I may - I might be wrong, so take it with a grain of salt:
it looks like operations executed for the 'lxc image copy' command is split into 2 parts: 
- client side. This retrieves instantly data from the simplestream, determines the real fingerprint for that alias reference (in this case - d6f281a2e523674bcd9822f3...), gives a copy command to the lxd, but does NOT give this fingerprint, but instead ONLY the alias, and then it adds the aliases based on its own calculated fingerprint
- server side: receives  an alias and not a real fingerprint, so it probably starts the same fingerprint/image selection algorithm. Probably, in my case, something was different with its data and it copied an image with a different fingerprint - the c3f79185497f one.  So things then went havoc.

Not sure what was different in my particular case, with the data, between lxd daemon and the lxc client.
Going further,  lxd seems to cache the simplestreams data for 1 hour. But lxc is always using fresh data. 
So, immediately after an image gets updated on the ubuntu simplestreams server (within 1 hour delta) - probably once every 1-2 months?, things can get into trouble, right?

Anyway - feel free to close this bug if you consider so. 
And thank you again for your software and time!

**Comment by cr1st1p** [comment]
it just  happened again...

[lxc_info.txt](https://github.com/lxc/lxd/files/3830865/lxc_info.txt)
[client.log](https://github.com/lxc/lxd/files/3830920/client.log)
[server.log](https://github.com/lxc/lxd/files/3830921/server.log)

file from inside lxd daemon - [simplestreams.yaml.gz](https://github.com/lxc/lxd/files/3830924/simplestreams.yaml.gz)

Looking at the simplestreams.xml, I see that the alias '18.04' is present, of course, twice,first one being the 'virtual-machine' (c3f... fingerprint) entry, and second one being the 'container' one (d6... fingerprint).

Checking a little bit the code for 'lxd' daemon.. why do I get the impression that the code for determining the exact  image and fingerprint to use is different than in the client code?
in client code I saw call to dereferenceAlias -> remote.GetImageAliasType -> GetImageAlias (which tries firs 'container' as a type, in case it is set to ''. 
While in lxd daemon I see only
```
                // Look for a matching alias
                for _, entry := range entry.Aliases {
                        if entry.Name != fp {
                                continue
                        }

                        if imageType != "" && entry.Type != imageType {
                                continue
                        }

                        fp = entry.Target
                        break
                }
```
I don't think I've seen the POST request to LXD to specify the 'container' image type, so probably the code from lxd daemon just stopped at first entry matching '18.04' - which was the wrong one.
Shouldn't the image type be also put in the request to LXD to copy the image?

**Comment by cr1st1p** [comment]
Continuing: based on my observation regarding the 1 hour cache inside lxd, while writing previous comment, I paused for 1 hour, then ran again the image copy command.
And... it worked.
The most important part, as I've observed in my last comment, is that now the order for alias '18.04' is different in the simplestream.yaml file - first is now 'container' type, and 2nd is 'virtual-machine'.
Either code to build that yaml file is time dependent (probably not), or it just reached a different ubuntu server, with a different data repartition.
[simplestreams.yaml.gz](https://github.com/lxc/lxd/files/3831104/simplestreams.yaml.gz)

So, my suggestion:
a) code for image selection algorithms should be as close as possible between lxc client and lxd daemon - now I don't think they are that close.
b) even if you do a), I recommend image type ('container' vs 'virtual-machine') to be passed from lxc client to lxd daemon for that image copy operation. if lxc client has chosen a particular image type, you really would want lxd daemon to do the same. 
And maybe even better/correct, to pass the real fingerprint in addition to the alias. My understanding is alias value is needed for automatic image updates, but current operation is on a very specific image fingerprint.

You can just use the uploaded simplestream.yaml file to replicate - upload the one from previous comment into  your lxd daemon.

**Comment by stgraber** [comment]
I think this got fixed in master as f59e9808ff8975e00bb5fefe350ecf9c06f7a61f
